### PR TITLE
On invalid prefix, throw a TypeError

### LINKED ISSUE #18058: Dependent function type with a wildcard argument crashes the compiler
## Compiler version

3.3.0

## Minimized code

```Scala
trait F:
  type A

type G = (f: _ <: F) => f.A
```

## Output

Throws a MatchError in the compiler.

Stack trace:

<details>

Crashes with stack trace,
```
scala.MatchError: dotty.tools.dotc.core.Types$UnspecifiedErrorType$@3db432c2 (of class dotty.tools.dotc.core.Types$UnspecifiedErrorType$) while running pickler on rs$line$1
Exception in thread "main" scala.MatchError: dotty.tools.dotc.core.Types$UnspecifiedErrorType$@3db432c2 (of class dotty.tools.dotc.core.Types$UnspecifiedErrorType$)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:295)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:162)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:626)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$20(TreePickler.scala:642)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:642)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:638)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:320)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:339)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:584)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:374)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:374)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:608)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:338)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:584)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:374)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:374)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:624)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:798)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:805)
        at dotty.tools.dotc.transform.Pickler.run$$anonfun$1$$anonfun$1(Pickler.scala:90)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:142)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:142)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:324)
        at scala.collection.immutable.List.map(List.scala:246)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:328)
        at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:150)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:247)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:263)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:271)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:280)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:280)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:201)
        at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:87)
        at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:307)
        at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:269)
        at dotty.tools.repl.ReplDriver.run$$anonfun$1(ReplDriver.scala:177)
        at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:211)
        at dotty.tools.repl.ReplDriver.runBody$$anonfun$1(ReplDriver.scala:185)
        at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80)
        at dotty.tools.repl.ReplDriver.runBody(ReplDriver.scala:185)
        at dotty.tools.repl.ReplDriver.run(ReplDriver.scala:178)
        at dotty.tools.repl.ReplDriver.interpretCommand(ReplDriver.scala:479)
        at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:276)
        at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:169)
        at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:172)
        at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:211)
        at dotty.tools.repl.ReplDriver.runBody$$anonfun$1(ReplDriver.scala:185)
        at dotty.tools.runner.ScalaClassLoader$.asContext(ScalaClassLoader.scala:80)
        at dotty.tools.repl.ReplDriver.runBody(ReplDriver.scala:185)
        at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:172)
        at dotty.tools.repl.ReplDriver.tryRunning(ReplDriver.scala:134)
        at dotty.tools.repl.Main$.main(Main.scala:7)
        at dotty.tools.repl.Main.main(Main.scala)
```
</details>

### PR DESCRIPTION
**Pull request by odersky**
Previously, the type was set to UnspecifiedErrorType. But that violates the rule that every ErrorType has to be reported. If that does not happen, crashes in pickler will result.

Fixes #18058

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -5400,6 +5401,9 @@ object Types {
     def explanation(using Context): String = msg.message
   }
 
+  /** Note: Make sure an errors is reported before construtcing this
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:5404]:
```suggestion
  /** Note: Make sure an error is reported before constructing this
```

**Review by dwijnand — COMMENTED** [review]
Should `type G = (f: _ <: F) => f.A` not be able to compile?

**Comment by odersky** [comment]
I don't think it should compile. `f` does not have a type we can use.

**Comment by dwijnand** [comment]
The test case goes through typedFunctionType's typedDependent, which encodes the function as a RefinedTypeTree, using an approximate AppliedTypeTree refined by the more specific param-dependant type.  For `f.A` we use `nonDependentResultApprox` to get the "least supertype of `resultType` that does not contain parameter dependencies".  Wouldn't F#A's supertype Any be that supertype?  My thinking is perhaps the issue is that `<: F` should be handled in the typemap's `derivedSelect` before failing in `withPrefix`...

Other that that, my change was to avoid a compiler crash caused by an assertion error.  I see now that using a TypeError makes the whole `typedFunctionType` fail and then associate the error message to the function tree, which looks good.

So I think perhaps this should compile, but not crashing in pickling is progress.

**Comment by odersky** [comment]
We could widen to `Any`, but that's no enough. The problem is we are writing a dependent function type, which is expanded to a generic type (that's where the avoid map was applied) and also an `apply` method. The `apply` method does not have a reasonable type either.

**Comment by clavigne** [comment]
Thanks for the prompt fix!

**Pull request merged**